File: python/app/vulns/path_traversal.py
"""Path Traversal — OWASP 2021 A01 Broken Access Control  ->  OWASP 2025 A01 / A05.

A path-traversal flaw lets an attacker step *outside* the directory an endpoint
was meant to serve (or write into) by smuggling ``..`` segments — or an absolute
path — into a filename that the code joins onto a trusted base and then hands to
``open()``. Unlike injection, no ORM/driver fixes it: safe file access is
application logic (canonicalize, then confirm the result stays inside the base).

The comment/annotation format here mirrors the reference module
``app/vulns/injection_sql.py``: every permutation carries a standard header with
its CWE, an exploitation-likelihood rating, prevalence notes, an example
request, and the corresponding safe pattern. Every dangerous sink is marked
``VULNERABLE:`` and the file ends with ONE clearly-labelled SAFE reference
handler.

The endpoints are supposed to be restricted to ``app/data/public`` (which holds
``welcome.txt``); the target the traversals reach is ``app/data/secret.txt``,
one level up.
"""
import os

from flask import Blueprint, request, jsonify, Response

bp = Blueprint("path_traversal", __name__, url_prefix="/path")

# The directory the file endpoints are *supposed* to be confined to, and the
# scratch directory the "archive extractor" is *supposed* to unpack into.
BASE_PUBLIC_DIR = os.path.join(os.path.dirname(__file__), "..", "data", "public")
EXTRACT_DIR = os.path.join(os.path.dirname(__file__), "..", "data", "extract")


# =============================================================================
# PERMUTATION 1 — Arbitrary file READ via unchecked path join
# OWASP 2021 A01 Broken Access Control (path traversal)  ->  OWASP 2025 A01 / A05
# CWE-22: Improper Limitation of a Pathname to a Restricted Directory ('Path Traversal')
# EXPLOITATION LIKELIHOOD: HIGH — pre-auth, deterministic, zero tooling. One
#   query parameter selects the file; a single '..' hop or an absolute path
#   reads anything the process can (app secrets, /etc/passwd, config, keys).
# PREVALENCE TODAY: greenfield MEDIUM (frameworks ship safe helpers like
#   send_from_directory/safe_join, but hand-rolled file-preview, avatar/image,
#   report and template-include endpoints still concatenate a user-named path)
#   | legacy/10yr tech-debt HIGH (download.php?file=... style code that opens
#   base + user_input with no normalization is endemic in old codebases).
# EXAMPLE:
#   curl 'http://127.0.0.1:5000/path/read?file=../secret.txt'   -> Flag: PY_TRAVERSAL_OK
#   curl 'http://127.0.0.1:5000/path/read?file=/etc/passwd'
#   curl 'http://127.0.0.1:5000/path/read?file=../../../../etc/passwd'
# FIX: realpath the joined path and verify it stays within the base dir, or use
#   the framework's safe_join / send_from_directory (see the SAFE handler below).
# =============================================================================
@bp.get("/read")
def read_file():
    filename = request.args.get("file", "welcome.txt")
    # VULNERABLE: user-controlled name joined onto the base dir with NO
    # normalization or containment check; '..' and absolute paths escape.
    path = os.path.join(BASE_PUBLIC_DIR, filename)
    try:
        with open(path) as fh:
            contents = fh.read()
    except OSError as exc:
        return jsonify({"file": filename, "path": path, "error": str(exc)}), 404
    return jsonify(
        {"file": filename, "resolved_path": os.path.realpath(path), "contents": contents}
    )


# =============================================================================
# PERMUTATION 2 — Arbitrary file DOWNLOAD via unchecked path join
# OWASP 2021 A01 Broken Access Control (path traversal)  ->  OWASP 2025 A01 / A05
# CWE-22: Improper Limitation of a Pathname to a Restricted Directory ('Path Traversal')
# EXPLOITATION LIKELIHOOD: HIGH — identical traversal to P1, just framed as a
#   file download: the same '..'/absolute-path payload streams any readable file
#   back with an attachment header. Pre-auth, deterministic, no tooling.
# PREVALENCE TODAY: greenfield MEDIUM ("download this attachment/export by name"
#   endpoints are routinely hand-rolled with a Content-Disposition header even in
#   modern apps, and the safe download helpers are under-used) | legacy/10yr
#   tech-debt HIGH (getFile?name=... download servlets/handlers everywhere).
# EXAMPLE:
#   curl -OJ 'http://127.0.0.1:5000/path/download?name=../secret.txt'
#   curl 'http://127.0.0.1:5000/path/download?name=/etc/passwd'
# FIX: canonicalize + containment-check (or send_from_directory), and derive the
#   download filename from an allow-list, never from the raw path.
# =============================================================================
@bp.get("/download")
def download():
    name = request.args.get("name", "welcome.txt")
    # VULNERABLE: same unchecked join; whatever path the client names is streamed
    # back as an attachment, including '..' escapes and absolute paths.
    path = os.path.join(BASE_PUBLIC_DIR, name)
    try:
        with open(path, "rb") as fh:
            data = fh.read()
    except OSError as exc:
        return jsonify({"name": name, "path": path, "error": str(exc)}), 404
    download_name = os.path.basename(name) or "download"
    resp = Response(data, mimetype="application/octet-stream")
    resp.headers["Content-Disposition"] = f'attachment; filename="{download_name}"'
    resp.headers["X-Resolved-Path"] = os.path.realpath(path)
    return resp


# =============================================================================
# PERMUTATION 3 — Arbitrary file WRITE via "archive extraction" (Zip Slip)
# OWASP 2021 A01 Broken Access Control (path traversal)  ->  OWASP 2025 A01 / A05
# CWE-22: Improper Limitation of a Pathname to a Restricted Directory ('Path Traversal')
# EXPLOITATION LIKELIHOOD: MEDIUM-HIGH — arbitrary WRITE is more powerful than a
#   read (overwrite web-root code, drop a cron/systemd unit or ~/.ssh key -> RCE),
#   but landing impact usually needs a known writable, executable target path and
#   sometimes a second step to trigger it, so it rates just below the direct read.
# PREVALENCE TODAY: greenfield MEDIUM (zipfile.extractall sanitizes members by
#   default, but hand-rolled extraction loops, tar handling, and upload-unpack /
#   storage-sync code that writes join(dir, entry.name) keep Zip Slip recurring
#   across every ecosystem — it was a broad 2018 disclosure wave) | legacy/10yr
#   tech-debt HIGH (manual per-entry extraction with no path check is the norm).
# EXAMPLE:
#   curl -X POST -H 'Content-Type: application/json' \
#     -d '{"entries":[{"name":"../../pwned.txt","content":"owned"}]}' \
#     http://127.0.0.1:5000/path/extract
#   -> writes app/pwned.txt, OUTSIDE the intended extract dir (absolute names work too)
# FIX: for each entry, realpath(join(dir, name)) and reject anything that does
#   not stay under the extraction root (same check as the SAFE handler below).
# =============================================================================
@bp.post("/extract")
def extract():
    body = request.get_json(silent=True) or {}
    entries = body.get("entries", [])
    os.makedirs(EXTRACT_DIR, exist_ok=True)
    results = []
    for entry in entries:
        name = entry.get("name", "")
        content = entry.get("content", "")
        # VULNERABLE: the archive entry name is joined onto the extract dir with
        # NO containment check; '../../pwned.txt' (or an absolute path) writes
        # OUTSIDE extractDir — the classic "Zip Slip" arbitrary file write.
        dest = os.path.join(EXTRACT_DIR, name)
        parent = os.path.dirname(dest)
        if parent:
            os.makedirs(parent, exist_ok=True)
        with open(dest, "w") as fh:
            fh.write(content)
        results.append({"name": name, "written_to": os.path.realpath(dest)})
    return jsonify(
        {"extract_dir": os.path.realpath(EXTRACT_DIR), "extracted": results}
    )


# =============================================================================
# SAFE REFERENCE — canonicalize, then enforce containment within the base dir.
# os.path.realpath resolves '..' segments and symlinks to a single absolute
# path; requiring it to equal the base or sit under base + os.sep rejects every
# traversal payload BEFORE the file is opened. This is the safe counterpart to
# PERMUTATION 1 (and the same check fixes P2's download and P3's per-entry
# write). Shown so the vulnerable/safe pair can be diffed by SAST and learners.
# =============================================================================
@bp.get("/read-safe")
def read_safe():
    filename = request.args.get("file", "welcome.txt")
    base = os.path.realpath(BASE_PUBLIC_DIR)
    requested = os.path.realpath(os.path.join(base, filename))
    # SAFE: reject anything whose canonical path escapes the intended base dir.
    if requested != base and not requested.startswith(base + os.sep):
        return jsonify(
            {"error": "path escapes base directory", "file": filename}
        ), 403
    try:
        with open(requested) as fh:
            contents = fh.read()
    except OSError as exc:
        return jsonify({"error": str(exc), "file": filename}), 404
    return jsonify({"file": filename, "resolved_path": requested, "contents": contents})

File: javascript/routes/path_traversal.js
/**
 * Path Traversal — OWASP 2021 A01 Broken Access Control | OWASP 2025 A01 / A05.
 *
 * Same annotation format as the reference SQL-injection module: each
 * permutation carries a standard header describing the flaw, its CWE, an
 * exploitation-likelihood rating, a prevalence estimate, an example request,
 * and the corresponding safe pattern. The dangerous sink on each handler is
 * flagged inline with `VULNERABLE:`.
 *
 * The file-serving handlers are anchored at BASE_PUBLIC (data/public/, holds
 * welcome.txt); the target the endpoints are NOT supposed to reach is
 * data/secret.txt one level up. The genuine dangerous primitive here is
 * `path.resolve(base, userInput)`: developers reach for it believing it
 * "cleans" a path, but it neither contains the result to `base` nor rejects
 * absolute inputs — an absolute argument REPLACES the base entirely, and `../`
 * segments climb above it.
 *
 * Sinks demonstrated (all genuine, all exploitable when the app runs):
 *   P1  fs.readFileSync(path.resolve(base, file))    -> arbitrary file READ    (CWE-22)
 *   P2  createReadStream(path.resolve(base, name))   -> arbitrary file DOWNLOAD (CWE-22)
 *   P3  fs.writeFileSync(path.join(dir, entry.name)) -> Zip Slip arbitrary WRITE (CWE-22)
 */
'use strict';

const express = require('express');
const fs = require('fs');
const path = require('path');

const router = express.Router();

// Directory the file endpoints are SUPPOSED to be restricted to, and the
// extraction target for the Zip-Slip demo. Both live under data/.
const BASE_PUBLIC = path.join(__dirname, '..', 'data', 'public');
const EXTRACT_DIR = path.join(BASE_PUBLIC, 'incoming');

// ===========================================================================
// PERMUTATION 1 — Arbitrary file read via path.resolve(base, userInput)
// OWASP 2021 A01 Broken Access Control (path traversal)  ->  OWASP 2025 A01 / A05
// CWE-22: Improper Limitation of a Pathname to a Restricted Directory ('Path Traversal')
//   (absolute-path variant is CWE-36 Absolute Path Traversal)
// EXPLOITATION LIKELIHOOD: HIGH — pre-auth, a single GET, fully deterministic,
//   no tooling. `?file=../secret.txt` escapes the public dir; `?file=/etc/passwd`
//   (absolute) makes resolve() discard the base outright; enough `../` climbs to
//   filesystem root. Classic "download report/template/avatar by name" bug.
// PREVALENCE TODAY: greenfield MEDIUM (static assets now go through hardened
//   middleware like express.static/send that reject `..`, and object storage
//   replaces filesystem reads — but every hand-rolled "serve this file by name"
//   route, plus misplaced faith in path.resolve/normalize as a sanitizer, keeps
//   reintroducing it; a very common AI-generated snippet) | legacy/10yr
//   tech-debt HIGH (file-by-name endpoints written before containment checks
//   were routine are everywhere in older codebases).
// EXAMPLE:
//   GET /path/read?file=welcome.txt                          -> intended file
//   GET /path/read?file=../secret.txt                        -> escapes -> data/secret.txt (flag)
//   GET /path/read?file=/etc/passwd                          -> absolute replaces base
//   GET /path/read?file=../../../../../../../../etc/passwd   -> climb to root then /etc/passwd
// FIX: resolve THEN verify containment — reject unless the resolved path equals
//   base or startsWith(base + path.sep); canonicalize with fs.realpathSync to
//   defeat symlinks, or better serve an allow-listed id -> filename map.
//   See /path/read-safe.
// ===========================================================================
router.get('/read', (req, res) => {
  const file = req.query.file || '';
  // VULNERABLE: user input resolved against the base with no containment check —
  // `../` escapes and an absolute path replaces BASE_PUBLIC entirely.
  const resolved = path.resolve(BASE_PUBLIC, file);
  try {
    const contents = fs.readFileSync(resolved, 'utf8');
    res.json({ base: BASE_PUBLIC, file, resolved, contents });
  } catch (err) {
    res.status(404).json({ base: BASE_PUBLIC, file, resolved, error: err.message });
  }
});

// ===========================================================================
// PERMUTATION 2 — Traversal in a file download (Content-Disposition attachment)
// OWASP 2021 A01 Broken Access Control (path traversal)  ->  OWASP 2025 A01 / A05
// CWE-22: Improper Limitation of a Pathname to a Restricted Directory ('Path Traversal')
//   (absolute-path variant is CWE-36 Absolute Path Traversal)
// EXPLOITATION LIKELIHOOD: HIGH — same primitive as P1, framed as a download:
//   pre-auth, one GET, deterministic. The attacker points `?name=` at any file
//   and the server streams it back as an attachment. Absolute paths accepted.
// PREVALENCE TODAY: greenfield MEDIUM (custom download endpoints that build a
//   Content-Disposition and stream a file chosen by a request parameter remain
//   common even in modern apps, since framework static handlers do not cover
//   "download this generated/user file by name") | legacy/10yr tech-debt HIGH
//   (report/invoice/export download servlets and CGI-era `?file=` handlers are a
//   canonical source of traversal in old code).
// EXAMPLE:
//   curl -OJ 'localhost:3000/path/download?name=../secret.txt'   # -> data/secret.txt (flag)
//   curl      'localhost:3000/path/download?name=/etc/passwd'    # -> absolute path
// FIX: resolve and enforce the same base-containment check before streaming, or
//   map an opaque id to a server-side filename; never pass request input to the
//   filesystem. See /path/read-safe for the containment pattern.
// ===========================================================================
router.get('/download', (req, res) => {
  const name = req.query.name || '';
  // VULNERABLE: download target resolved from user input with no containment —
  // traversal and absolute paths stream any file the process can read.
  const resolved = path.resolve(BASE_PUBLIC, name);
  res.setHeader('Content-Disposition', `attachment; filename="${path.basename(resolved)}"`);
  res.setHeader('X-Resolved-Path', resolved); // echo the effective path (like the SQL demo echoes the query)
  const stream = fs.createReadStream(resolved);
  stream.on('error', (err) => {
    if (!res.headersSent) res.status(404).json({ base: BASE_PUBLIC, name, resolved, error: err.message });
  });
  stream.pipe(res);
});

// ===========================================================================
// PERMUTATION 3 — Zip Slip: arbitrary file WRITE on archive extraction
// OWASP 2021 A01 Broken Access Control (path traversal)  ->  OWASP 2025 A01 / A05
// CWE-22: Improper Limitation of a Pathname to a Restricted Directory ('Path Traversal')
// EXPLOITATION LIKELIHOOD: MEDIUM-HIGH — the write itself is trivial and
//   deterministic: an entry named `../../pwned.txt` lands OUTSIDE the extraction
//   directory. Escalation to RCE is high-impact but contingent on a useful write
//   target (web root, cron.d, a startup/rc file, .ssh/authorized_keys), which is
//   what makes it MEDIUM-HIGH rather than a guaranteed HIGH.
// PREVALENCE TODAY: greenfield LOW (mainstream archive libraries added Zip-Slip
//   guards after the 2018 Zip-Slip disclosure — modern unzip/tar reject `..` by
//   default — so new code that uses them is usually safe) | legacy/10yr
//   tech-debt MEDIUM (hand-rolled "loop the entries and write each to
//   destDir + entry.name" extractors, and pre-2018 library versions, still ship
//   in older upload/import/plugin pipelines).
// EXAMPLE:
//   curl -s localhost:3000/path/extract -H 'Content-Type: application/json' \
//     -d '{"entries":[{"name":"notes/ok.txt","content":"hi"},
//                     {"name":"../../pwned.txt","content":"owned"}]}'
//   -> the second entry resolves outside EXTRACT_DIR (…/javascript/data/pwned.txt)
// FIX: for every entry, resolve the target and reject it unless it stays within
//   the extraction dir (resolved === dir || startsWith(dir + path.sep)); drop
//   absolute names and any path containing `..`. See /path/extract-safe.
// ===========================================================================
router.post('/extract', (req, res) => {
  const entries = (req.body && req.body.entries) || [];
  const results = [];
  try {
    fs.mkdirSync(EXTRACT_DIR, { recursive: true });
    for (const entry of entries) {
      const name = (entry && entry.name) || '';
      const content = entry && entry.content != null ? String(entry.content) : '';
      // VULNERABLE: entry name joined onto the extraction dir with no containment
      // — a `../` in the name writes the file OUTSIDE EXTRACT_DIR (Zip Slip).
      const target = path.join(EXTRACT_DIR, name);
      const written = path.resolve(target);
      fs.mkdirSync(path.dirname(written), { recursive: true });
      fs.writeFileSync(written, content);
      const escaped = written !== path.resolve(EXTRACT_DIR) &&
        !written.startsWith(path.resolve(EXTRACT_DIR) + path.sep);
      results.push({ name, written, escaped });
    }
    res.json({ extractDir: path.resolve(EXTRACT_DIR), results });
  } catch (err) {
    res.status(500).json({ extractDir: path.resolve(EXTRACT_DIR), error: err.message, results });
  }
});

// ===========================================================================
// SAFE REFERENCE — canonicalize the resolved path and verify it stays within
// the intended base directory. Shown so the vulnerable/safe pair can be diffed
// by SAST tooling and learners. Mirrors os.path.realpath+startswith (Python) /
// Path.toRealPath().startsWith (Java): resolve, containment-check the string,
// then realpath the result to defeat symlink escapes before any read.
// ===========================================================================
router.get('/read-safe', (req, res) => {
  const file = req.query.file || '';
  const baseReal = fs.realpathSync(BASE_PUBLIC); // canonical base
  const resolved = path.resolve(baseReal, file);
  // SAFE: reject anything that does not stay under the base directory.
  if (resolved !== baseReal && !resolved.startsWith(baseReal + path.sep)) {
    return res.status(403).json({ file, resolved, error: 'path escapes base directory' });
  }
  try {
    const real = fs.realpathSync(resolved); // SAFE: canonicalize to defeat symlink escapes
    if (real !== baseReal && !real.startsWith(baseReal + path.sep)) {
      return res.status(403).json({ file, resolved: real, error: 'symlink escapes base directory' });
    }
    res.json({ file, resolved: real, contents: fs.readFileSync(real, 'utf8') });
  } catch (err) {
    res.status(404).json({ file, resolved, error: err.message });
  }
});

module.exports = { base: '/path', router };
